aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -87,9 +89,73 @@ impl AptosVM {
     pub fn internals(&self) -> AptosVMInternals {
         AptosVMInternals::new(&self.0)
     }
-
+    fn is_valid_for_constant(typ: &Type) -> bool {
+        use move_vm_types::loaded_data::runtime_types::Type::*;
+        match typ {
+            Bool | U8 | U64 | U128 | Address => true,
+            Vector(inner) => AptosVM::is_valid_for_constant(&(*inner)),
+            Signer
+            | Struct(_)
+            | StructInstantiation(_, _)
+            | Reference(_)
+            | MutableReference(_)
+            | TyParam(_) => false,
+        }
+    }
+    /// validation and generate args for entry function
+    /// validation includes:
+    /// 1. return signature is empty
+    /// 2. number of signers is same as the number of senders
+    /// 3. check args are no resource after signers
+    ///
+    /// after validation, add senders and non-signer arguments to generate the final args
+    fn validate_combine_signer_and_txn_args(
+        senders: Vec<AccountAddress>,
+        args: Vec<Vec<u8>>,
+        func: &LoadedFunctionInstantiation,
+    ) -> Result<Vec<Vec<u8>>, VMStatus> {
+        // entry function should not return
+        if !func.return_.is_empty() {
+            return Err(VMStatus::Error(StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE));
+        }
+        let mut n_signer_params = 0;
+        let mut is_signer_end = false;
+        for (_i, ty) in func.parameters.iter().enumerate() {
+            if matches!(ty, Type::Signer) {
+                n_signer_params += 1;
+            } else {
+                is_signer_end = true;
+            }
+            // check if the type contain resource type
+            if is_signer_end && !AptosVM::is_valid_for_constant(ty) {
+                return Err(VMStatus::Error(StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE));
+            }
+        }
+        if (n_signer_params + args.len()) != func.parameters.len() {
+            return Err(VMStatus::Error(StatusCode::NUMBER_OF_ARGUMENTS_MISMATCH));
+        }
+        let res = if n_signer_params == 0 {
+            args
+        } else {
+            if senders.len() != n_signer_params {
+                return Err(VMStatus::Error(
+                    StatusCode::NUMBER_OF_SIGNER_ARGUMENTS_MISMATCH,
+                ));
+            }
+            senders
+                .into_iter()
+                .map(|s| MoveValue::Signer(s).simple_serialize().unwrap())
+                .chain(args)
+                .collect()
+        };
```

Comment:
This makes sense, but lets put a few comments through this function so it's a little more straightforward what's going on.